File: solutions/swift/circular-buffer/1/Sources/CircularBuffer/CircularBuffer.swift
enum CircularBufferError: Error { case bufferEmpty, bufferFull }

struct CircularBuffer {
  var capacity: Int
  private var buffer: [Int]

  init(capacity: Int) {
    self.capacity = capacity
    self.buffer = [Int]()
  }

  mutating func clear() { buffer.removeAll() }

  mutating func overwrite(_ value: Int) {
    if buffer.count >= capacity { buffer.removeFirst() }
    try! write(value)
  }

  mutating func write(_ value: Int) throws {
    guard buffer.count < capacity else { throw CircularBufferError.bufferFull }
    buffer.append(value)
  }

  mutating func read() throws -> Int {
    guard !buffer.isEmpty else { throw CircularBufferError.bufferEmpty }
    return buffer.removeFirst()
  }
}

File: solutions/ruby/circular-buffer/2/circular_buffer.rb
# frozen_string_literal: true

# class CircularBuffer
class CircularBuffer
  class BufferEmptyException < StandardError; end
  class BufferFullException < StandardError; end

  def initialize(capacity) = (@capacity = capacity
                              @buffer = [])

  def full? = @buffer.size == @capacity
  def clear = @buffer.clear
  def read = @buffer.empty? ? raise(BufferEmptyException) : @buffer.shift
  def write(val) = full? ? raise(BufferFullException) : @buffer << val

  def write!(val) = (read if full?
                     write(val))
end
